File: 0684-redundant-connection/0684-redundant-connection.py
class Solution:
    def findRedundantConnection(self, edges: List[List[int]]) -> List[int]:
        n=len(edges)
        parent = [i for i in range(n+1)]

        def find(node):

            while node!=parent[node]:
                node=parent[node]

            return node 

        def union(i,j):
            pi = find(i)
            pj = find(j)

            parent[pj]=pi

        for s,d in edges:
            if find(s)== find(d):
                return [s,d]
            else:
                union(s,d)

File: 0684-redundant-connection/0684-redundant-connection.java
class Solution {
    
    int[] parent;
    
    public int find(int node){
        while(node != parent[node]){
            node = parent[node];
        }
        return node;
    }
    
    public void union(int i, int j){
        int pi = find(i);
        int pj = find(j);
        parent[pj] = pi;
    }
    
    public int[] findRedundantConnection(int[][] edges) {
        int n = edges.length;
        parent = new int[n+1];
        
        for(int i=0; i<parent.length; i++){
            parent[i] = i;
        }
        
        for(int[] edge: edges){
            if(find(edge[0]) == find(edge[1])){
                return edge;
            }else{
                union(edge[0],edge[1]);
            }
        }
        return edges[0];
    }
}